dsref/resolve.go

```diff
@@ -0,0 +1,137 @@
+package dsref
+
+import (
+	"context"
+	"errors"
+	"fmt"
+)
+
+var (
+	// ErrNotFound must be returned by a ref resolver that cannot resolve a given
+	// reference
+	ErrNotFound = errors.New("reference not found")
+)
+
+// Resolver finds the identifier and HEAD path for a dataset reference
+type Resolver interface {
+	// ResolveRef uses ref as an outParam, setting ref.ID and ref.Path on success
+	// some implementations of name resolution may make network calls
+	// the returned "source" value should be either an empty string, indicating
+	// the source was resolved locally, or the multiaddress of the network
+	// address that performed the resolution
+	ResolveRef(ctx context.Context, ref *Ref) (source string, err error)
+}
+
+// MemResolver holds maps that can do a cheap version of dataset resolution,
```

Comment:
Feels like this should remain in its own source file "mem_resolve.go". A lot of our codebase follows that pattern, interface in one file and mem-version in another, and I think it's a good thing.